Add type declarations where backwards-compatible

greg0ire opened the pull request:
This includes:
- private methods
- return type declarations of final protected methods
- return type declarations of public and protected methods of final classes

Parameter type declarations are a more delicate matter and should
probably be handled separately to make it easier to catch issues during
code review.

Type declarations can be more trusted than simple phpdoc when it
comes to static analysis, having them means we can infer the phpdoc of
calling methods with confidence.

Note that it seems that some of the phpdoc I initially inferred these
declarations from were apparently wrong, in particular some mentioning
`Doctrine\Dbal\Statement` when was is really passed around is
`Doctrine\Dbal\Driver\Statement`.

Closes #8538

simonberger commented on lib/Doctrine/ORM/Internal/Hydration/ArrayHydrator.php:
```diff
@@ -261,13 +261,14 @@ protected function hydrateRowData(array $row, array &$result)
-    private function updateResultPointer(array &$coll, $index, $dqlAlias, $oneToOne)
-    {
+    private function updateResultPointer(
+        array &$coll,
```
`null` is handled in the code for parameter `$coll`.

simonberger commented on lib/Doctrine/ORM/PersistentCollection.php:
```diff
@@ -292,17 +277,15 @@ static function ($a, $b): int {
      * @psalm-return array<string, mixed>
      */
-    public function getMapping()
+    public function getMapping(): array
```
Result can be null. Property var attribute needs a change as well.

greg0ire commented:
Thanks for the careful review @simonberger , much appreciated 👍

BackEndTea commented on lib/Doctrine/ORM/Cache/DefaultQueryCache.php:
```diff
@@ -344,11 +344,9 @@ public function put(QueryCacheKey $key, ResultSetMapping $rsm, $result, array $h
      * @param array<string,mixed> $assoc
      * @param mixed               $assocValue
      *
-     * @return mixed[]|null
```
I'd keep the annotation here. That at least makes it clear the return type wasn't forgotten here

ostrolucky commented on lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php:
```diff
@@ -98,13 +98,7 @@ public function update($entity)
-     * @return bool
-     */
-    private function updateCache($entity, $isChanged)
+    private function updateCache(object $entity, bool $isChanged): ?bool
```
rather than changing it from docblock version, i would prefer fixing the code here: `$isChanged = $isChanged || $cached;`

ostrolucky commented on lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php:
```diff
@@ -273,13 +272,11 @@ private function getEntity(array $data, $dqlAlias)
     }
 
     /**
-     * @param string $className
```
isn't this a class-string? should be marked as such, if so

ostrolucky commented on lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:
```diff
@@ -384,12 +382,8 @@ private function addDefaultDiscriminatorMap(ClassMetadata $class)
     /**
      * Gets the lower-case short name of a class.
-     *
-     * @param string $className
```
class-string?

ostrolucky commented on lib/Doctrine/ORM/Mapping/Driver/DatabaseDriver.php:
```diff
@@ -526,12 +522,8 @@ private function getTablePrimaryKeys(Table $table)
-     *
-     * @param string $tableName
-     *
-     * @return string
```
class-string?

ostrolucky commented on lib/Doctrine/ORM/Mapping/Driver/XmlDriver.php:
```diff
@@ -821,11 +819,9 @@ private function columnToArray(SimpleXMLElement $fieldMapping)
     /**
      * Parse / Normalize the cache configuration
      *
-     * @return mixed[]
```
Sometimes you remove these, sometimes don't, like in yamldriver or PersistentCollection. So which way is it, what's the policy?

greg0ire replied:
It's unclear to me what's best, will give it more thought later.

ostrolucky replied:
If I may, I slightly prefer keeping them. There might be tools that don't respect psalm vendor specific annotations and mixed[] makes it clear to communicate we didn't forget to specify the type.

greg0ire replied:
Ok, I'll restore them then 👍

ostrolucky commented on lib/Doctrine/ORM/NativeQuery.php:
```diff
@@ -60,7 +60,7 @@ public function getSQL()
 
     /**
-     * {@inheritdoc}
+     * {@inheritDoc}
```
what was your motivation to change these? I fuzzily remember phpcs understood lowercase variant only

greg0ire replied:
I don't remember, and we seem to be using a mix of both in the project :sweat: .

Will look into it later.

ostrolucky commented on lib/Doctrine/ORM/PersistentCollection.php:
```diff
@@ -143,18 +141,13 @@ public function setOwner($entity, array $assoc)
-     * @return Mapping\ClassMetadata
-     */
-    public function getTypeClass()
+    public function getTypeClass(): Mapping\ClassMetadata
```
This one is dangerous. We don't enforce type in constructor and there also is ClassMetadataInfo

ostrolucky commented on lib/Doctrine/ORM/PersistentCollection.php:
```diff
@@ -348,20 +329,16 @@ public function setDirty($dirty)
-     * @return object
      */
-    public function remove($key)
+    public function remove($key): object
```
should be `?object`. It calls parent::remove, ArrayCollection returns null in some cases and interface also says T|null.

please also take care of references to this method, like offsetUnset

ostrolucky commented on lib/Doctrine/ORM/Query.php:
```diff
@@ -279,7 +277,7 @@ private function parse()
-     * {@inheritdoc}
+     * {@inheritDoc}
      */
     protected function _doExecute()
```
return type here should probably be added

greg0ire replied:
That one causes so many test failures :scream:

ostrolucky replied:
So is it actually `mixed`?

greg0ire replied:
Indeed, it's not just the tests, the sql executor hierarchy has inconsistent return type declarations. `MultiTableDeleteExecutor` declares `int`, so does `SingleTableDeleteUpdateExecutor`.

ostrolucky commented on lib/Doctrine/ORM/Query.php:
```diff
@@ -523,10 +521,8 @@ public function setQueryCacheLifetime($timeToLive): self
-     * @return int
      */
-    public function getQueryCacheLifetime()
+    public function getQueryCacheLifetime(): int
```
Not sure why SA isn't complaining here. `queryCacheTTL` prop is uninitializd so I think it may return null

greg0ire replied:
I think that's because we are at a dangerously lenient level right now.

ostrolucky commented on lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php:
```diff
@@ -223,15 +225,13 @@ private function getColumnAlias($columnName, $mode, array $customRenameColumns)
      *
      * This depends on the renaming mode selected by the user.
      *
-     * @param string $className
```
class-string?

ostrolucky commented on lib/Doctrine/ORM/Query/ResultSetMappingBuilder.php:
```diff
@@ -223,15 +225,13 @@ private function getColumnAlias($columnName, $mode, array $customRenameColumns)
      * @psalm-return array<array-key, string>
      */
-    private function getColumnAliasMap($className, $mode, array $customRenameColumns)
+    private function getColumnAliasMap(string $className, $mode, array $customRenameColumns): array
```
`int $mode`? Either that or docblock should be fixed, shouldn't it?

ostrolucky commented on lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:
```diff
@@ -95,24 +95,16 @@ public function generateEntityRepositoryClass($fullClassName)
     /**
      * Generates the namespace, if class do not have namespace, return empty string instead.
-     *
-     * @param string $fullClassName
```
class-string

ostrolucky commented on lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:
```diff
@@ -95,24 +95,16 @@ public function generateEntityRepositoryClass($fullClassName)
     /**
      * Generates the class name
-     *
-     * @param string $fullClassName
```
class-string

ostrolucky commented on lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:
```diff
@@ -129,22 +121,15 @@ private function generateClassName($fullClassName)
      * Generates the namespace statement, if class do not have namespace, return empty string instead.
      *
      * @param string $fullClassName The full repository class name.
```
class-string

ostrolucky commented on lib/Doctrine/ORM/Tools/EntityRepositoryGenerator.php:
```diff
@@ -129,22 +121,15 @@ private function generateClassName($fullClassName)
     }
 
-    /**
-     * @param string $fullClassName
```
class-string

ostrolucky commented on lib/Doctrine/ORM/Tools/Pagination/Paginator.php:
```diff
@@ -219,10 +208,8 @@ private function useOutputWalker(Query $query)
     /**
      * Appends a custom tree walker to the tree walkers hint.
-     *
-     * @param string $walkerClass
```
class-string

ostrolucky commented on lib/Doctrine/ORM/Tools/ResolveTargetEntityListener.php:
```diff
@@ -110,12 +109,9 @@ public function loadClassMetadata(LoadClassMetadataEventArgs $args)
+     * @param mixed[] $mapping
      */
-    private function remapAssociation($classMetadata, $mapping)
+    private function remapAssociation(ClassMetadata $classMetadata, $mapping): void
```
any reason why not declaring type for $mapping?

greg0ire replied:
That one slipped through :P

ostrolucky reviewed (commented):
One comment remaining. Also, you are adding object type declarations, but these are being removed in https://github.com/doctrine/orm/pull/8613. That will cause mess of a conflict.

greg0ire commented on lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:
```diff
@@ -24,9 +24,9 @@
+use Doctrine\DBAL\Driver\ResultStatement as DriverStatement;
 use Doctrine\DBAL\LockMode;
 use Doctrine\DBAL\Platforms\AbstractPlatform;
-use Doctrine\DBAL\Statement;
```
Unsure if we should wait for https://github.com/doctrine/dbal/pull/4596 and revert this change in use statements. I think that change might be good because more forward compatible.

ostrolucky replied:
I'll leave you decide. But to me, such change seems a bit unrelated to this PR

greg0ire commented:
I will wait until #8613 is resolved

greg0ire committed:
Add type declarations where backwards-compatible

This includes:
- private methods
- return type declarations of final protected methods
- return type declarations of public and protected methods of final classes

Parameter type declarations are a more delicate matter and should
probably be handled separately to make it easier to catch issues during
code review.

Type declarations can be more trusted than simple phpdoc when it
comes to static analysis, having them means we can infer the phpdoc of
calling methods with confidence.

Note that it seems that some of the phpdoc I initially inferred these
declarations from were apparently wrong, in particular some mentioning
Doctrine\Dbal\Statement when was is really passed around is
Doctrine\Dbal\Driver\Statement.

greg0ire commented:
The 7.1 PR has been merged, let's rebase to be 100% sure it still works
